Lib/test/test_import/__init__.py

```diff
@@ -1894,6 +1894,30 @@ def test_absolute_circular_submodule(self):
             str(cm.exception),
         )
 
+    def test_singlephase_circular(self):
+        """Regression test for gh-123950
+
+        Import a single-phase-init module that imports itself
+        from the PyInit_* function (before it's added to sys.modules).
+        Manages its own cache (which is `static`, and so incompatible
+        with multiple interpreters or interpreter reset).
+        """
+        name = '_testsinglephase_circular'
+        filename = _testsinglephase.__file__
+        loader = importlib.machinery.ExtensionFileLoader(name, filename)
+        spec = importlib.util.spec_from_file_location(name, filename,
+                                                      loader=loader)
+        mod = importlib._bootstrap._load(spec)
```

Comment:
I don't know if I love using this API for the test since it's somewhat of a hack to start, but I also understand not wanting to paste in 2 more lines of boilerplate to get the module created and set `sys.modules`.

Reply:
It was done elsewhere in the file too; I merged the usage into a common helper. If it breaks, there's now just one place to change.

(We never got around to adding proper public API for multiple modules in one shared library, but then, it's really only useful for testing import mechanism...)

In this case I slightly prefer *not* to touch `sys.modules` here -- the “inner” import should do that.